src/digraph-cyclicity.jl

```diff
@@ -0,0 +1,320 @@
+    * blockedmap: mapping of the blocking / unblocking consequences
+* `vmap`: vector map containing the link from the old to the new nodes of the directed graph
+* `cycle`: storage of the channel
+* startnode = v: optional argument giving the starting node. In the first iteration,
+the same as v, otherwise it should be passed.
+
+# Returns
+* done: tells whether a circuit has been found in the current exploration.
+"""
+function circuit_iter end
+@traitfn function circuit_iter{T<:Integer}(v::T, dg::::IsDirected, vis::JohnsonVisitor{T}, 
+vmap::Vector{T}, cycle::Channel, startnode::T = v)
+    done = false
+    push!(vis.stack, v)
+    vis.blocked[v] = true
+    for w in out_neighbors(dg, v)
+        if w == startnode
+            put!(cycle, vmap[vis.stack])
+            done = true
+        elseif !vis.blocked[w]
+            done = circuit_iter(w, dg, vis, vmap, cycle, startnode)
+        end
+    end
+    if done
+        unblock!(v, vis.blocked, vis.blockedmap)
+    else
+        for w in out_neighbors(dg, v)
+            if !in(vis.blockedmap[w], v)
+                push!(vis.blockedmap[w], v)
+            end
+        end
+    end
+    pop!(vis.stack)
+    return done
+end
+
+
+"""
+    itercycles(dg::DiGraph, cycle::Channel)
+
+Compute all cycles of the given directed graph, using
+[Johnson's algorithm](http://epubs.siam.org/doi/abs/10.1137/0204007).
+
+Iterative version of the algorithm, using Channels to stop the exploration
+after a given number of cycles.
+"""
+function itercycles end
+@traitfn function itercycles(dg::::IsDirected, cycle::Channel)
+    sccs = strongly_connected_components(dg)
+    for scc in sccs
+        while length(scc) > 1
+            wdg, vmap = induced_subgraph(dg, scc)
+            shift!(scc)
+            visitor = JohnsonVisitor(wdg)
+            circuit_iter(1, wdg, visitor, vmap, cycle)
+        end
+    end
+end
+
+@doc_str """
+    simplecyclescount(dg::DiGraph, ceiling = 10^6)
+
+Count the number of cycles in a directed graph, using
+[Johnson's algorithm](http://epubs.siam.org/doi/abs/10.1137/0204007).
+
+The `ceiling` is here to avoir memory overload if there are a lot of cycles in the graph.
+Default value is 10^6, but it can be higher or lower. You can use the function
+```maxsimplecycles(dg::DiGraph, byscc::Bool = true)``` to get an idea of the
+theoretical maximum number or cycles.
+
+It returns `len`, the number of cycles if below the ceiling, the ceiling otherwise.
+"""
+function simplecyclescount end
+@traitfn function simplecyclescount(dg::::IsDirected, ceiling = 10^6)
+    len = 0
+    for cycle in Iterators.take(Channel(c->itercycles(dg,c)), ceiling)
+        len += 1
+    end
+    return len
+end
+
+@doc_str """
+    simplecycles_iter(dg::DiGraph, ceiling = 10^6)
+
+Search all cycles of the given directed graph, using
+[Johnson's algorithm](http://epubs.siam.org/doi/abs/10.1137/0204007),
+up to the ceiling (avoid memory overload).
+
+If the graph is small, the ceiling will not bite and

```

Comment:
It means that the default ceiling will not be reached. I will change the sentence to make that clear.